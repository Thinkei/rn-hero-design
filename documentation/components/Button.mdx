---
name: Button
menu: Components
---

import { Playground } from 'docz'
import { Props } from 'gatsby-theme-docz/src/components/Props';
import { Button } from 'rn-hero-design'
import { RED } from 'rn-hero-design/src/themes/hero/variables';
import { IPhoneX } from './DeviceFrame'

# Button

## Basic usage

<Playground>
  <IPhoneX screenStyle={{ padding: 16, paddingTop: 48 }}>
    <Button
      text="Filled Button"
      onPress={() => {}}
      wrapperStyle={{ marginBottom: 16 }}
    />
    <Button
      text="Loading Filled Button"
      onPress={() => {}}
      loading
      wrapperStyle={{ marginBottom: 16 }}
    />
    <Button
      text="Disabled Filled Button"
      onPress={() => {}}
      disabled
      wrapperStyle={{ marginBottom: 32 }}
    />
    <Button
      variant="outlined"
      text="Outlined Button"
      onPress={() => {}}
      wrapperStyle={{ marginBottom: 16 }}
    />
    <Button
      variant="outlined"
      text="Loading Outlined Button"
      onPress={() => {}}
      loading
      wrapperStyle={{ marginBottom: 16 }}
    />
    <Button
      variant="outlined"
      text="Disabled Outlined Button"
      onPress={() => {}}
      disabled
      wrapperStyle={{ marginBottom: 32 }}
    />
    <Button
      text="Custom Button"
      onPress={() => {}}
      wrapperStyle={{ marginBottom: 16, backgroundColor: RED, borderRadius: 8 }}
    />
  </IPhoneX>
</Playground>

## Properties

<Props props_={{
  text: { type: 'String', required: true, description: 'Display text on button' },
  onPress: { type: 'event => void', required: true },
  loading: { type: 'Boolean', required: false, defaultValue: { value: 'false' } },
  disabled: { type: 'Boolean', required: false, defaultValue: { value: 'false' } },
  variant: { type: '"filled" | "outlined"', required: false },
  wrapperStyle: { type: 'Object', required: false },
  textStyle: { type: 'Object', required: false },
  theme: { type: 'Object', required: false, description: 'The theme object contains style of all components. Can be injected by `injectTheme` HOC' },
}} />

