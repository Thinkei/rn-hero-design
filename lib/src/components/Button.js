import React from 'react';
import {
  View,
  Text,
  TouchableOpacity,
  ActivityIndicator,
  StyleSheet,
} from 'react-native';
import t from 'prop-types';
import heroTheme from '../themes/hero';

const Button = ({
  text,
  onPress,
  disabled,
  loading,
  icon,
  rightIcon,
  style,
  theme = heroTheme,
}) => {
  const Wrapper = loading || disabled ? View : TouchableOpacity;
  return (
    <Wrapper
      onPress={onPress}
      style={StyleSheet.flatten([
        theme.button.wrapper,
        loading ? theme.button.loading : null,
        disabled ? theme.button.disabled : null,
        style,
      ])}
    >
      {loading ? (
        <ActivityIndicator size="small" color="white" />
      ) : (
        <Text style={theme.button.text}>{text}</Text>
      )}
    </Wrapper>
  );
};

Button.propTypes = {
  /**
   * The text displayed inside button
   */
  text: t.string.isRequired,
};

export default Button;
