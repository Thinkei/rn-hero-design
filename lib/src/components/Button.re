open ReactNative;

[@bs.module "../helpers/injectTheme.js"]
external injectTheme: React.component('a) => React.component('a) = "default";

module ThemeComponent = {
  let theme = Hero_Theme.default;
  let make = (~render) => {
    render(theme);
  };
};

type variant =
  | Filled
  | Outlined;

let getStylesByVariant = (variant: variant, styles: Js.t('a)) => {
  switch (variant) {
  | Filled => {
      "wrapper": styles##filledWrapper,
      "text": styles##filledText,
      "loadingIndicator": styles##filledLoadingIndicator,
    }
  | Outlined => {
      "wrapper": styles##outlinedWrapper,
      "text": styles##outlinedText,
      "loadingIndicator": styles##outlinedLoadingIndicator,
    }
  };
};

[@react.component]
let make =
    (
      ~text,
      ~onPress,
      ~loading=false,
      ~disabled=false,
      ~variant=Filled,
      ~theme=Hero_Theme.default,
      ~wrapperStyle=Style.style(),
      ~textStyle=Style.style(),
    ) => {
  let styles = getStylesByVariant(variant, theme##button);

  module Wrapper = {
    let style =
      StyleSheet.flatten([|
        theme##button##wrapper,
        styles##wrapper,
        disabled ? theme##button##disabledWrapper : Style.style(),
        wrapperStyle,
      |]);

    [@react.component]
    let make = (~children): React.element =>
      loading || disabled
        ? <View style> children </View>
        : <TouchableOpacity onPress style> children </TouchableOpacity>;
  };

  <Wrapper>
    {loading && !disabled
       ? <ActivityIndicator size=ActivityIndicator.Size.small />
       : <Text
           style={StyleSheet.flatten([|
             theme##button##text,
             styles##text,
             disabled ? theme##button##disabledWrapper : Style.style(),
             textStyle,
           |])}>
           text->React.string
         </Text>}
  </Wrapper>;
};

/* ThemeComponent.make(~render=theme => */
/*   <Wrapper> */
/*     <Text> {(text ++ string_of_int(theme))->React.string} </Text> */
/*   </Wrapper> */
/* ); */

let default = make;
