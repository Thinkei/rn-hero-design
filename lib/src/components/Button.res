open ReactNative

type variant = [#filled | #outlined]

module ButtonVariant: {
  type t
  @inline("filled")
  let filled: t
  @inline("outlined")
  let outlined: t
} = {
  type t = string
  @inline
  let filled = "filled"
  @inline
  let outlined = "outlined"
}

@get external getColorProperty: Style.t => Color.t = "color"

let getStylesByVariant = (variant: ButtonVariant.t, styles: 'a) => {
  open ButtonVariant
  switch variant {
  | v if v === outlined => {
      "wrapper": styles["outlinedWrapper"],
      "text": styles["outlinedText"],
      "loadingIndicator": styles["outlinedLoadingIndicator"],
    }
  | _ => {
      "wrapper": styles["filledWrapper"],
      "text": styles["filledText"],
      "loadingIndicator": styles["filledLoadingIndicator"],
    }
  }
}

@react.component
let make = (
  ~testID=?,
  ~text,
  ~onPress,
  ~loading=false,
  ~disabled=false,
  ~variant: ButtonVariant.t,
  ~theme=Hero_Theme.default,
  ~wrapperStyle=Style.style(),
  ~textStyle=Style.style(),
) => {
  let styles = variant->getStylesByVariant(theme["button"])

  module Wrapper = {
    let style = StyleSheet.flatten([
      theme["button"]["wrapper"],
      styles["wrapper"],
      disabled ? theme["button"]["disabledWrapper"] : Style.style(),
      wrapperStyle,
    ])

    @react.component
    let make = (~children): React.element =>
      loading || disabled
        ? <View ?testID style> children </View>
        : <TouchableOpacity ?testID activeOpacity=0.2 onPress style> children </TouchableOpacity>
  }

  <Wrapper>
    {loading && !disabled
      ? <ActivityIndicator
          size=ActivityIndicator.Size.small color={getColorProperty(styles["loadingIndicator"])}
        />
      : <Text
          style={StyleSheet.flatten([
            theme["button"]["text"],
            styles["text"],
            disabled ? theme["button"]["disabledText"] : Style.style(),
            textStyle,
          ])}>
          {text->React.string}
        </Text>}
  </Wrapper>
}

let default = Helpers.injectTheme(make)
