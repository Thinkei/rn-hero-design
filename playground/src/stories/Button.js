import React from 'react';
import { Alert, ScrollView } from 'react-native';
import { Button, injectTheme } from 'rn-hero-design';

const callback = () => Alert.alert('Button clicked!');

const ButtonScreen = ({ theme }) => (
  <ScrollView contentContainerStyle={{ padding: theme.variables.MEDIUM_SIZE }}>
    <Button
      text="Filled Button"
      onPress={callback}
      wrapperStyle={{ marginBottom: theme.variables.MEDIUM_SIZE }}
    />
    <Button
      text="Loading Filled Button"
      onPress={callback}
      loading
      wrapperStyle={{ marginBottom: theme.variables.MEDIUM_SIZE }}
    />
    <Button
      text="Disabled Filled Button"
      onPress={callback}
      disabled
      wrapperStyle={{ marginBottom: theme.variables.MEDIUM_SIZE }}
    />

    <Button
      variant="outlined"
      text="Outlined Button"
      onPress={callback}
      wrapperStyle={{ marginBottom: theme.variables.MEDIUM_SIZE }}
    />
    <Button
      variant="outlined"
      text="Loading Outlined Button"
      onPress={callback}
      loading
      wrapperStyle={{ marginBottom: theme.variables.MEDIUM_SIZE }}
    />
    <Button
      variant="outlined"
      text="Disabled Outlined Button"
      onPress={callback}
      disabled
      wrapperStyle={{ marginBottom: theme.variables.MEDIUM_SIZE }}
    />

    <Button
      text="Custom Button"
      onPress={callback}
      wrapperStyle={{
        marginBottom: theme.variables.MEDIUM_SIZE,
        backgroundColor: 'tomato',
        borderRadius: theme.variables.SMALL_SIZE,
      }}
    />
  </ScrollView>
);

export default injectTheme(ButtonScreen);
