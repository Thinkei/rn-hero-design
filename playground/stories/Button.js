import React from 'react';
import { View, Alert, StyleSheet } from 'react-native';
import { Button } from 'rn-hero-design';

const callback = () => Alert.alert('Button clicked!');

const ButtonScreen = () => (
  <View style={styles.container}>
    <Button text="Button" onPress={callback} style={{ marginBottom: 16 }} />
    <Button
      text="Loading"
      onPress={callback}
      loading
      style={{ marginBottom: 16 }}
    />
    <Button
      text="Disabled"
      onPress={callback}
      disabled
      style={{ marginBottom: 16 }}
    />
  </View>
);

const styles = StyleSheet.create({
  container: {
    padding: 16,
  },
});

export default ButtonScreen;
