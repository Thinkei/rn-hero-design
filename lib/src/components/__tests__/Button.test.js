import React from 'react';
import Button from '../Button.gen';
import { fireEvent, render } from '@testing-library/react-native';

describe('Button', () => {
  it.each`
    element
    ${<Button text="Button" onPress={() => {}} />}
    ${<Button text="Button" onPress={() => {}} disabled />}
    ${<Button text="Button" onPress={() => {}} loading />}
    ${<Button text="Button" onPress={() => {}} variant="outlined" />}
  `('renders correctly', async ({ element }) => {
    const { baseElement } = render(element);
    expect(baseElement).toMatchSnapshot();
  });

  it('should feedback when users press', async () => {
    const callback = jest.fn();
    const { getByText } = render(<Button text="Tap here" onPress={callback} />);

    const button = getByText('Tap here');
    fireEvent.press(button);
    expect(callback).toBeCalledTimes(1);
  });

  it('should not feedback when disabled', async () => {
    const callback = jest.fn();
    const { getByText } = render(
      <Button disabled text="Tap here" onPress={callback} />,
    );

    const button = getByText('Tap here');
    fireEvent.press(button);
    expect(callback).toBeCalledTimes(0);
  });
});
