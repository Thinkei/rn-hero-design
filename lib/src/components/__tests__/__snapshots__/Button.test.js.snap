// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#1dbeee",
      "borderRadius": 26.25,
      "display": "flex",
      "height": 52.5,
      "justifyContent": "center",
      "opacity": 1,
      "paddingHorizontal": 17.5,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#fcfcfc",
        "fontFamily": undefined,
        "fontSize": 17.5,
        "lineHeight": 17.5,
      }
    }
  >
    Button
  </Text>
</View>
`;

exports[`Button renders correctly 2`] = `
<View
  onPress={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#d9dbdf",
      "borderRadius": 26.25,
      "borderWidth": 0,
      "display": "flex",
      "height": 52.5,
      "justifyContent": "center",
      "paddingHorizontal": 17.5,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#fcfcfc",
        "fontFamily": undefined,
        "fontSize": 17.5,
        "lineHeight": 17.5,
      }
    }
  >
    Button
  </Text>
</View>
`;

exports[`Button renders correctly 3`] = `
<View
  onPress={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#1dbeee",
      "borderRadius": 26.25,
      "display": "flex",
      "height": 52.5,
      "justifyContent": "center",
      "paddingHorizontal": 17.5,
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="#fcfcfc"
    hidesWhenStopped={true}
    size="small"
  />
</View>
`;

exports[`Button renders correctly 4`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "transparent",
      "borderColor": "#1dbeee",
      "borderRadius": 26.25,
      "borderWidth": 1.2,
      "display": "flex",
      "height": 52.5,
      "justifyContent": "center",
      "opacity": 1,
      "paddingHorizontal": 17.5,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#1dbeee",
        "fontFamily": undefined,
        "fontSize": 17.5,
        "fontWeight": "500",
        "lineHeight": 17.5,
      }
    }
  >
    Button
  </Text>
</View>
`;
